### PR TITLE
feat: improve attribute/quantity change handling

### LINKED ISSUE #10372: [Łukasz] Add field to Configuration model indicating attrubute change or value quantity change and make sure its filled.
Add field to Configuration model indicating attribute change or value quantity change. 
(Coulde be flags on attribute and / or value level)

### LINKED ISSUE #10652: Quantity Input (all bundle Ui types) - not able to enter 2 (or more) digits amount
This issue occurs quantity inputs of all bundle types (dropdownProduct, radioGroupProduct and checkBoxListproduct). 

**repro:**
- configure 'Configurable Camera Bundle'
- go to group 'Accessories'
- add any 'Tripod'
- click into quantity input and try to type in eg: '25'

**Issue**: this results typically in value '2', because a request is fired as soon as you type and you loose the focus.

ideally there should be some delay before firing the change event, so that you can enter a also several digits.

similar issue happens, if i use backspace to delete the current input. before i can enter a new quantity the whole stepper disappears, as  item is considered deselected.

### LINKED ISSUE #10655: CheckBoxListProduct quantity is reset to 1 after deselecting and re-selecting
**repro**
- configure 'Configurable Camera Bundle'
- go to group 'Accessories'
- add any tripod
- increase quantity e.g.: '2'
- remove the Tripod
- re-add the same tripod

**issue**: Quantity is still '2'.
**expected**: quantity shall be 1

see also: https://4prsat.axshare.com/#id=68anbn&p=bundling_multi-select_qtybox&g=1

`If a user removes a bundle item that has a quantity entered greater than '1' and subsequently RE-selects (restores) the same bundle item, the default value of '1' shall appear in the quantity box (A). The previously entered quantity value is not restored.`

### LINKED ISSUE #10709: Qty for simple checkbox: '0' should deselect
1) According to wireframes a selected checkbox should be deselected if value is set to '0':
Screenshot 2021-01-14 at 16 49 45

This is not working: Decrease to 0 not possible. Although manually entering 0 will lead to 1.

### PR DESCRIPTION
**Pull request by luklus**
close #10372

### CODE REVIEW

File: feature-libs/product-configurator/rulebased/components/attribute/types/checkbox/configurator-attribute-checkbox.component.ts [hunk 1]
```diff
@@ -27,6 +27,7 @@ export class ConfiguratorAttributeCheckBoxComponent
   attributeCheckBoxForm = new FormControl('');
 
   ngOnInit() {
+    console.log(this.attribute, 'checkbox');
```
Comment (kacperknapik) [inline feature-libs/product-configurator/rulebased/components/attribute/types/checkbox/configurator-attribute-checkbox.component.ts]: remove `console.log`

File: feature-libs/product-configurator/rulebased/components/attribute/types/multi-selection-bundle/configurator-attribute-multi-selection-bundle.component.html [hunk 2]
```diff
@@ -10,7 +10,8 @@
     *ngFor="let value of attribute.values"
     [product]="value"
     [multiSelect]="true"
-    [disabledAction]="disableDeselectAction$ | async"
+    [disabledAction]="false"
```
Comment (kacperknapik) [inline feature-libs/product-configurator/rulebased/components/attribute/types/multi-selection-bundle/configurator-attribute-multi-selection-bundle.component.html]: are we explicitly setting this to `false`? or is it a mistake

File: feature-libs/product-configurator/rulebased/components/attribute/types/radio-button/configurator-attribute-radio-button.component.html [hunk 3]
```diff
@@ -1,13 +1,13 @@
 <fieldset>
   <legend style="display: none">{{ attribute.label }}</legend>
   <div id="{{ createAttributeIdForConfigurator(attribute) }}">
-    <div class="form-quantity">
+    <div class="form-quantity" *ngIf="withQuantity">
       <label>{{ 'configurator.attribute.quantity' | cxTranslate }}</label>
       <cx-item-counter
-        [allowZero]="!attribute.selectedSingleValue"
+        [allowZero]="true"
```
Comment (kacperknapik) [inline feature-libs/product-configurator/rulebased/components/attribute/types/radio-button/configurator-attribute-radio-button.component.html]: are we explicitly setting this to `true`? or is it a mistake

**Review by kacperknapik — CHANGES_REQUESTED** [review]

**Comment by kacperknapik** [comment]
check travis aswell

File: feature-libs/product-configurator/rulebased/components/attribute/types/checkbox/configurator-attribute-checkbox.component.ts (hunk 1 above)
Comment (luklus) [reply feature-libs/product-configurator/rulebased/components/attribute/types/checkbox/configurator-attribute-checkbox.component.ts:30]: it is GH issue :) in code we don't have a console log 
Zrzut ekranu 2021-01-13 090757

File: feature-libs/product-configurator/rulebased/components/attribute/types/multi-selection-bundle/configurator-attribute-multi-selection-bundle.component.html (hunk 2 above)
Comment (luklus) [reply feature-libs/product-configurator/rulebased/components/attribute/types/multi-selection-bundle/configurator-attribute-multi-selection-bundle.component.html]: it is no mistake, we dont need more this setting, i will remove all code related to it

File: feature-libs/product-configurator/rulebased/components/attribute/types/radio-button/configurator-attribute-radio-button.component.html (hunk 3 above)
Comment (luklus) [reply feature-libs/product-configurator/rulebased/components/attribute/types/radio-button/configurator-attribute-radio-button.component.html]: yes, they should all be able to minus to 0. when quantity equal 0 item is removed from selection

File: feature-libs/product-configurator/rulebased/components/attribute/types/checkbox/configurator-attribute-checkbox.component.ts (hunk 1 above)
Comment (luklus) [reply feature-libs/product-configurator/rulebased/components/attribute/types/checkbox/configurator-attribute-checkbox.component.ts]: fixed

**Review by luklus — COMMENTED** [review]
> check travis aswell

fixed

**Review by Uli-Tiger — CHANGES_REQUESTED** [review]
@luklus  There are quite some duplicated blocks according to sonar:
image
maybe it makes sense to move the common logic into an own class and write unit test for them once. That would also help with the code coverage, which is low especially for the duplicated blocks

**Comment by Uli-Tiger** [comment]
@luklus : as this blocks a lot of testing tasks, i will merge it. But create another task to improve coverage a bit. Usually we aim for overage > 90%
follow-up: https://github.com/SAP/spartacus/issues/10713

**Pull request merged**